packages/ui/src/forum/hooks/useForumCategoryDetails.ts

```diff
@@ -0,0 +1,81 @@
+import { useEffect, useState } from 'react'
+
+import {
+  useGetForumPostsCountLazyQuery,
+  useGetForumPostsLazyQuery,
+  useGetForumSubCategoriesQuery,
+  useGetForumThreadsCountQuery,
+  useGetForumThreadsQuery,
+} from '@/forum/queries'
+import {
+  asForumPost,
+  asForumThread,
+  asSubCategory,
+  ForumCategory,
+  ForumPost,
+  ForumSubCategory,
+  ForumThread,
+} from '@/forum/types'
+
+export interface CategoryListItemProps {
+  category: ForumCategory
+}
+
+interface CategoryDetails {
+  subcategories?: ForumSubCategory[]
+  threadCount?: number
+  topThread?: ForumThread & { postCount?: number }
+  latestPost?: ForumPost
+}
+
+export const useForumCategoryDetails = (categoryId: string): CategoryDetails => {
```

Comment:
I see that there's a lot going on which is hard for me to process as some fetch/query dependencies are intervened. This makes me think that it might be better to split this hook and rewrite dependent components a bit.

*   this hook should fetch a minimal set of "core" data
*   A non-core data (that can be fetched independently) are
    *   total threads
    *   last post
    *   popular thread
    *   moderators?

I think that the `ForumCatregoryListItem` should just pass `forumCategoryId` to the other components.

```tsx
const ForumCategoryListItem = (category) => {
  const { threadCount } = useThreadCound(category.id)
  return (
    <>
      // ... Other components like title, subcategories, etc
      <ThreadCount count={threadCount} /> // this might be fetched here
      <LastPost forumCategoryId={category.id} />
      <TopThread forumCategoryId={category.id} />
    </>
  )
}
```

:point\_up:  The above is just a general idea, not the real code to be implemented.

Reply:
In here the `<LastPost />` should fetch its own data using `useLastPost()` hook.

Reply:
Additionally, I'm worried that every time the optional data is fetched, and the `details` is updated the whole hook make the component using it to re-render while it could be limited to the component that uses that data. That's a minor issue, though as the child components are using memo anyway, but still, a thing to keep in mind.

ps.: The above claim might not be 100% accurate. I don't recall now if the main component will be re-rendered when children are rendered. :thinking:

Reply:
I think the previous code was indeed triggering more unnecessary updates. In any case these changes do make thing more readable :+1: